### PR TITLE
Various commits chopped off from PR 1881

### PR DESCRIPTION
**Pull request by ferencdg**
PR 1881 stuck in a permanent review state, so I chopped off some commits that I need for other PRs. Comments for those commits are on the PR 1881 page. Please merge, so I can continue creating other PRs, otherwise I am **blocked**.

### CODE REVIEW

File: source/agora/common/Amount.d [hunk 1]
```diff
@@ -260,11 +260,27 @@ public struct Amount
     ***************************************************************************/
 
     pragma(inline, true)
-    public void percentage (ubyte percentage)
+    public bool percentage (ushort pct)
```
Comment (Geod24) [inline source/agora/common/Amount.d:263]: Need to update the UT:
```
source/agora/common/Amount.d:499:    amt.percentage(33);
source/agora/common/Amount.d:501:    amt.percentage(33);
source/agora/common/Amount.d:503:    amt.percentage(10);
```

Also commit message: "percentage", not "percantage".

File: source/agora/common/Amount.d [hunk 2]
```diff
@@ -260,11 +260,27 @@ public struct Amount
...
     {
-        assert(percentage <= 100);
-        this.value *= percentage;
+        this.value *= pct;
```
Comment (Geod24) [inline source/agora/common/Amount.d]: There should be a validity check to avoid overflow which would turn an invalid value into a valid one.

File: source/agora/common/Amount.d (hunk 1 above)
Comment (Geod24) [inline source/agora/common/Amount.d:263]: Why did you change the parameter name ?

File: source/agora/common/Amount.d [hunk 3]
```diff
@@ -260,11 +260,27 @@ public struct Amount
...
         this.value /= 100;
+        return isValid();
+    }
+
+    /***************************************************************************
+
+        Take a percentage of the `value` and assert if the result is invalid
+
+        Params:
+            percentage = Requested percentage
```
Comment (Geod24) [inline source/agora/common/Amount.d]: Mismatch with the parameter name

File: source/agora/utils/TxBuilder.d [hunk 4]
```diff
@@ -424,6 +424,25 @@ public struct TxBuilder
...
+    ***************************************************************************/
+
+    public ref typeof(this) deductRemaining ()
+        return
+    {
+        this.leftover.value = Amount(0);
+
+        return this;
+    }
+
```
Comment (Geod24) [inline source/agora/utils/TxBuilder.d]: Since fee seems to be the only use case, I think it'd be better to have something that reference it, such as `addFee`. Also this function is lacking attributes.

File: source/agora/common/Amount.d [hunk 5]
```diff
@@ -47,6 +47,12 @@ public struct Amount
     /// Internal data storage
     private ulong value;
 
+    /// The size of the Amount object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
+    {
+        return value.sizeof;
+    }
+
```
Comment (Geod24) [inline source/agora/common/Amount.d]: Just call sizeof from the caller ?

File: source/agora/consensus/Fee.d [hunk 6]
```diff
@@ -427,17 +456,19 @@ public class FeeManager
...
-                assert(peekUTXO(input.utxo, utxo));
+                if (!peekUTXO(input.utxo, utxo))
+                    return "unable to find input for utxo";
                 tot_in.mustAdd(utxo.output.value);
             }
 
-            assert(tx.getSumOutput(tot_out), "Not validated block in" ~
-                "getTXSetFees");
+            if (!tx.getSumOutput(tot_out))
+                return "Not validated block in getTXSetFees";
```
Comment (Geod24) [inline source/agora/consensus/Fee.d]: That message doesn't make much sense, think we can reword it ?

File: source/agora/node/Ledger.d [hunk 7]
```diff
@@ -669,8 +669,9 @@ public class Ledger
         in uint[] missing_validators) nothrow @safe
     {
         Amount tot_fee, tot_data_fee;
-        this.fee_man.getTXSetFees(tx_set, &this.utxo_set.peekUTXO, tot_fee,
-            tot_data_fee);
+        auto fee_res = this.fee_man.getTXSetFees(tx_set, &this.utxo_set.peekUTXO, tot_fee, tot_data_fee);
+        if (fee_res != null)
+            assert(0, fee_res);
```
Comment (Geod24) [inline source/agora/node/Ledger.d]:
```D
if (auto fee_res = ....)
    assert(0, fee_res);
```

File: source/agora/consensus/Fee.d [hunk 8]
```diff
@@ -427,17 +456,19 @@ public class FeeManager
             foreach (input; tx.inputs)
             {
                 UTXO utxo;
-                assert(peekUTXO(input.utxo, utxo));
+                if (!peekUTXO(input.utxo, utxo))
+                    return "unable to find input for utxo";
```
Comment (Geod24) [inline source/agora/consensus/Fee.d]: Other messages are capitalized.

File: source/agora/consensus/Fee.d [hunk 9]
```diff
@@ -401,6 +401,33 @@ public class FeeManager
...
+        Returns: string describing the error, if an error happened, null otherwise
+
+    ***************************************************************************/
+
+    public string getAdjustedTXFee (in Transaction tx, scope UTXOFinder peekUTXO,
+        out Amount tot_fee) nothrow @safe
+    {
+        Amount tot_data_fee;
+
+        auto res = getTXSetFees([tx], peekUTXO, tot_fee, tot_data_fee);
```
Comment (Geod24) [inline source/agora/consensus/Fee.d]: `getTXSetFees` should really accept a range, or you could pass `(&tx)[0 .. 1]` to avoid the needless allocation.
But let's fix this later as it might create other issues with scoping and `@safe`.

File: source/agora/common/Amount.d (hunk 1 above)
Comment (linked0) [inline source/agora/common/Amount.d:263]: The return value is only used in `mustPercentage`. Why don't you call the `isValid` with `assert`?

**Review by Geod24 — CHANGES_REQUESTED** [review]
The `getSizeInBytes` doesn't seem very documented and is not tested. Since it will dictate the fees in the protocol, it'd be nice to know the expectation. Additionally, I think we should limit it to the Transaction module to avoid contaminating all data structures.

File: source/agora/common/Amount.d [hunk 10]
```diff
@@ -260,11 +260,27 @@ public struct Amount
...
+        Take a percentage of the `value` and assert if the result is invalid
+
+        Params:
+            percentage = Requested percentage
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public Amount mustPercentage (ushort pct)
```
Comment (linked0) [inline source/agora/common/Amount.d:285]: Another name like `getPercentage would be better.

File: source/agora/consensus/data/Transaction.d [hunk 11]
```diff
@@ -60,6 +60,17 @@ public struct Transaction
     /// The data to store
     public DataPayload payload;
 
+    /// The size of the Transaction object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
+    {
+        ulong size = type.sizeof + payload.sizeInBytes();
+        foreach (const ref input; inputs)
```
Comment (linked0) [inline source/agora/consensus/data/Transaction.d]: Suggested change.
`this.inputs`

File: source/agora/consensus/data/Transaction.d [hunk 12]
```diff
@@ -60,6 +60,17 @@ public struct Transaction
...
+            size += input.sizeInBytes();
+        foreach (const ref output; outputs)
```
Comment (linked0) [inline source/agora/consensus/data/Transaction.d]: Suggested change.
`this.outputs`

File: source/agora/consensus/Fee.d (hunk 6 above)
Comment (ferencdg) [reply source/agora/consensus/Fee.d]: I can rename it, it came from a previous commit though

File: source/agora/consensus/data/Transaction.d [hunk 13]
```diff
@@ -122,6 +133,12 @@ public struct Output
     /// The lock condition for this Output
     public Lock lock;
 
+    /// The size of the Output object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
+    {
+        return value.sizeInBytes() + lock.sizeInBytes();
```
Comment (linked0) [inline source/agora/consensus/data/Transaction.d]: Shouldn't we add `this` to the member variables?

File: source/agora/consensus/data/Transaction.d [hunk 14]
```diff
@@ -183,6 +200,12 @@ public struct Input
     /// included in the block. Use for implementing relative time locks.
     public uint unlock_age = 0;
 
+    /// The size of the Input object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
+    {
+        return unlock.sizeInBytes() + unlock_age.sizeof + utxo.sizeof;
```
Comment (linked0) [inline source/agora/consensus/data/Transaction.d]: Ditto

File: source/agora/script/Lock.d [hunk 15]
```diff
@@ -45,13 +45,25 @@ public struct Lock
 
     /// May either be a Hash, or a sequence of opcodes
     public const(ubyte)[] bytes;
+
+    /// The size of the Lock object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
+    {
+        return type.sizeof + bytes.length * bytes[0].sizeof;
```
Comment (linked0) [inline source/agora/script/Lock.d]: Ditto

File: source/agora/script/Lock.d [hunk 16]
```diff
@@ -45,13 +45,25 @@ public struct Lock
...
 /// Contains a data tuple or a set of push opcodes
 public struct Unlock
 {
     /// May be: <signature>, <signature, key>, <push opcodes>
     public const(ubyte)[] bytes;
+
+    /// The size of the Unlock object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
+    {
+        return bytes.length * bytes[0].sizeof;
```
Comment (linked0) [inline source/agora/script/Lock.d]: Ditto

File: source/agora/consensus/Fee.d [hunk 17]
```diff
@@ -401,6 +401,33 @@ public class FeeManager
...
+            tx = transaction for which we want to calculate the adjusted fee
+            peekUTXO = UTXO finder
+            tot_fee = adjusted fee
+
+        Returns: string describing the error, if an error happened, null otherwise
+
+    ***************************************************************************/
+
+    public string getAdjustedTXFee (in Transaction tx, scope UTXOFinder peekUTXO,
+        out Amount tot_fee) nothrow @safe
```
Comment (linked0) [inline source/agora/consensus/Fee.d:421]: Out of curiosity, what does the `tot` stand for?

**Review by linked0 — COMMENTED** [review]
Good work and some comments.

File: source/agora/common/Amount.d (hunk 1 above)
Comment (ferencdg) [reply source/agora/common/Amount.d:263]: because now we have a method called 
```D
    pragma(inline, true)
    public Amount mustPercentage (ushort pct)
    {
        percentage(pct) || assert(0);
        return this;
    }
```

and this method calls the percentage method, so I had to change the percenatege parameter to pct to be consistent

File: source/agora/common/Amount.d (hunk 10 above)
Comment (ferencdg) [reply source/agora/common/Amount.d:285]: I wanted to be consisten with mustAdd/mustSub

File: source/agora/consensus/data/Transaction.d (hunk 11 above)
Comment (ferencdg) [reply source/agora/consensus/data/Transaction.d]: kind of contraversial :) If we modify the the style guide that we have to prefix every class variable with this then I am happy to comply. VSCode uses different colors for class members vs local variables though.

File: source/agora/consensus/Fee.d (hunk 17 above)
Comment (ferencdg) [reply source/agora/consensus/Fee.d:421]: total -> same name used in getTXFee

File: source/agora/common/Amount.d (hunk 1 above)
Comment (ferencdg) [reply source/agora/common/Amount.d:263]: percentage and mustPercantage methods can be called by anyone and they behave differently:
percentage - returns false if invalid
mustPercantage - asserts if invalid

hence the `must` prefix

this is the same behaviour as with mustSub/sub and mustAdd/add

**Comment by ferencdg** [comment]
> The `getSizeInBytes` doesn't seem very documented and is not tested. 

what else could I say?
```D
    /// The size of the Transaction object
    public ulong sizeInBytes() const nothrow pure @safe @nogc
```
> and is not tested. 

if someone adds a new field to the Transaction object, and forget to change the sizeInBytes method, then I will not be able to catch that with unit tests. If someone removes a field, and forget the change the sizeInBytes method, he will receive a compile error. I can add a unit test that calculates the size and then I canculate it on paper and compare... But I already have done that.


> Additionally, I think we should limit it to the Transaction module to avoid contaminating all data structures.

do you mean use `package` modifier on the sizeInBytes() method? I am going to use that method in Fee for example and that is not in the same package

File: source/agora/utils/TxBuilder.d (hunk 4 above)
Comment (ferencdg) [reply source/agora/utils/TxBuilder.d]: We already have an addFee method called: 'deduct', but tho use it you need to know how much value is left in the output. That makes using 'deduct' method inconvinient in some cases(like when using it with filters/maps...). I can call it deductRemaining, if that is okay, or I can remove both methods to addFee and addRemainingAsFee

File: source/agora/consensus/Fee.d (hunk 9 above)
Comment (ferencdg) [reply source/agora/consensus/Fee.d]: Just using a fixed size array with length 1 would avoid allocations(on the heap).

**Comment by ferencdg** [comment]
thanks for the review @Geod24, @linked0 I fixed everything, and still have 1 pending question: https://github.com/bosagora/agora/pull/1955#issuecomment-820846559

if you are okay with my answer, then this could be merged

**Comment by Geod24** [comment]
This triggers a (valid) assert:
```
#### FATAL ERROR: 
This node was started at source/agora/test/Flash.d:528
This most likely means that the node crashed due to an uncaught exception
If not, please file a bug at https://github.com/Geod24/localrest/
Full error: core.exception.AssertError@source/agora/common/Amount.d(341)
----------------
runtime.d:815 [0x563497768a4b]
runtime.d:774 [0x563497768273]
dmain2.d:292 [0x56349777bf00]
deh.d:46 [0x5634977b226f]
dwarfeh.d:332 [0x56349777cb02]
/home/runner/work/agora/agora/source/agora/test/Base.d:140 [0x563496f089b1]
exception.d:430 [0x56349776773f]
exception.d:595 [0x563497767c24]
/home/runner/work/agora/agora/source/agora/common/Amount.d:341 [0x563496e06413]
/home/runner/work/agora/agora/source/agora/consensus/Fee.d:329 [0x5634973bcfdc]
/opt/hostedtoolcache/dc/ldc2-1.25.0/x64/ldc2-1.25.0-linux-x86_64/bin/../import/std/algorithm/iteration.d:1026 [0x5634973bac9c]
/home/runner/work/agora/agora/source/agora/consensus/Fee.d:329 [0x563496fdc189]
/home/runner/work/agora/agora/source/agora/node/Ledger.d:620 [0x563496eb000e]
/home/runner/work/agora/agora/source/agora/node/Ledger.d:1173 [0x563496f0aef3]
/home/runner/work/agora/agora/source/agora/test/Base.d:478 [0x563496e63651]
```

**Comment by ferencdg** [comment]
> This triggers a (valid) assert:
> 
> ```
> #### FATAL ERROR: 
> This node was started at source/agora/test/Flash.d:528
> This most likely means that the node crashed due to an uncaught exception
> If not, please file a bug at https://github.com/Geod24/localrest/
> Full error: core.exception.AssertError@source/agora/common/Amount.d(341)
> ----------------
> runtime.d:815 [0x563497768a4b]
> runtime.d:774 [0x563497768273]
> dmain2.d:292 [0x56349777bf00]
> deh.d:46 [0x5634977b226f]
> dwarfeh.d:332 [0x56349777cb02]
> /home/runner/work/agora/agora/source/agora/test/Base.d:140 [0x563496f089b1]
> exception.d:430 [0x56349776773f]
> exception.d:595 [0x563497767c24]
> /home/runner/work/agora/agora/source/agora/common/Amount.d:341 [0x563496e06413]
> /home/runner/work/agora/agora/source/agora/consensus/Fee.d:329 [0x5634973bcfdc]
> /opt/hostedtoolcache/dc/ldc2-1.25.0/x64/ldc2-1.25.0-linux-x86_64/bin/../import/std/algorithm/iteration.d:1026 [0x5634973bac9c]
> /home/runner/work/agora/agora/source/agora/consensus/Fee.d:329 [0x563496fdc189]
> /home/runner/work/agora/agora/source/agora/node/Ledger.d:620 [0x563496eb000e]
> /home/runner/work/agora/agora/source/agora/node/Ledger.d:1173 [0x563496f0aef3]
> /home/runner/work/agora/agora/source/agora/test/Base.d:478 [0x563496e63651]
> ```

fixed, anything else needed before this PR can be merged? It is blocking #1917

File: source/agora/common/Amount.d (hunk 1 above)
Comment (Geod24) [reply source/agora/common/Amount.d:263]: Well `this.percentage(percentage)` would have worked but I see your point now.

File: source/agora/consensus/data/Transaction.d (hunk 11 above)
Comment (Geod24) [reply source/agora/consensus/data/Transaction.d]: It's not in the style guide ? Because we've always done it. Indeed should add it.

File: source/agora/consensus/data/DataPayload.d [hunk 18]
```diff
@@ -26,6 +26,11 @@ public struct DataPayload
     /// The byte array of transaction data
     public const(ubyte)[] data;
 
+    /// The size of the data DataPayload object
+    public ulong sizeInBytes() const nothrow pure @safe @nogc
```
Comment (Geod24) [inline source/agora/consensus/data/DataPayload.d]: Also missing the space before parens in all those functions

File: source/agora/consensus/Fee.d [hunk 19]
```diff
@@ -427,17 +457,19 @@ public class FeeManager
...
-                assert(peekUTXO(input.utxo, utxo));
+                if (!peekUTXO(input.utxo, utxo))
+                    return "Unable to find input for utxo";
                 tot_in.mustAdd(utxo.output.value);
             }
 
-            assert(tx.getSumOutput(tot_out), "Not validated block in" ~
-                "getTXSetFees");
+            if (!tx.getSumOutput(tot_out))
+                return "Transaction output value is invalid";
```
Comment (Geod24) [inline source/agora/consensus/Fee.d:466]: If those get triggered, isn't it logic error ?

Comment (ferencdg) [reply source/agora/consensus/Fee.d:466]: this function can get called (and is called in PR #1881) with partially verified transactions (transactions that got relayed to us by an attacker through putTransaction). It can very well happen that those transactions didn't have their input in the UTXO or their total sum doesn't add up, so we shouldn't assert

File: source/agora/consensus/data/Transaction.d (hunk 11 above)
Comment (ferencdg) [reply source/agora/consensus/data/Transaction.d]: > It's not in the style guide ? Because we've always done it. Indeed should add it.

Yeah, I just checked it is not there. I will go through this PR and add 'this' and also change the style guide

**Comment by linked0** [comment]
The last commit should be fixuped to the fourth one, I think.

**Review by Geod24 — APPROVED** [review]
Needs a rebase but g2g

**Comment by ferencdg** [comment]
> Needs a rebase but g2g

rebased and also prefixed everything with 'this'...
